src/web/debugging.md

```diff
@@ -3,247 +3,172 @@ title: Debugging Dart web apps
+   and displays the source code for your app's main file.
 
-1. In the console, try viewing a property of the component.
+1. Set a breakpoint inside a timer or event handler
+   by clicking to the left of one of its lines of code.
    <br>
+   For example, click the line number for the first line inside
+   an event handler or timer callback.
 
-   For example, try to view the `items` object:
-
-   * Enter `items`.
-     It doesn't work because JavaScript requires a `this.` prefix.
-
-   * Enter `this.items`.
-     This works because the JavaScript object has the same name as
-     the Dart object.
-     Thanks to custom formatters, you can see the Dart type and value.
-
-   * Enter `this.items[0]`.
-     This works because Dart lists map to JavaScript arrays.
-
-   * Enter `this.items.first`.
-     This doesn't work, because unlike the [Dart List class,][List]
-     JavaScript arrays don't have a `first` property.
-
-{% comment %}
-Dart DevTools:
-items.first
-items.isEmpty
-items.length
-items.runtimeType
-[PENDING: once indexing works in DDT, tell them to try items[0])
-{% endcomment %}
-
-1. Try [other DevTools features.][JavaScript debugging reference]
-
-
-## Changing DevTools settings
-
-This section covers settings that you might need to change
-as you debug your app.
-
-
-### Enabling custom formatters {#enabling-custom-formatters}
-
-To see Dart types in Chrome DevTools, you need to enable custom formatters.
+1. Trigger the event that causes the function call.
+   Execution stops at the breakpoint.
 
-1. From the DevTools **Customize and control DevTools ⋮** menu,
-   choose **Settings**. Or press <kbd>F1</kbd>.
-1. Select **Enable custom formatters**,
-   which is under the **Console** heading in the **Preferences** settings.
-1. Close Settings: Press <kbd>Esc</kbd> or click the **x** at the upper right.
+1. In the **Variables** pane, inspect the values of variables.
 
-{% comment %}
-The best description I found was https://www.mattzeunert.com/2016/02/19/custom-chrome-devtools-object-formatters.html.
-{% endcomment %}
+1. Resume script execution, and trigger the event again or press **Pause**.
+   Execution pauses again.
 
+1. Try stepping through code line-by-line using the
+   **Step in**, **Step over**, and **Step out** buttons.
 
-### Disabling and re-enabling source maps {#enabling-and-disabling-source-maps}
+   {{site.alert.note}}
+     Dart DevTools doesn't step into SDK code.
+     For example, if you press **Step in** at a call to `print()`,
+     you go to the next line, not into the SDK code that implements `print()`.
+   {{site.alert.end}}
 
-Chrome DevTools <!-- and Dart DevTools --> enables source maps, by default.
-You might want to temporarily disable source maps so that you can see
-the generated JavaScript code.
+1. Change your source code and reload the Chrome window that's running the app.
+   The app quickly rebuilds and reloads.
+   **[PENDING: mention something about losing your breakpoints?]**
```

Comment:
Correct. @bkonyi is working on properly resolving this issue. We have a temporary workaround that is not being leveraged by `webdev` or the `Flutter` CLI. @devoncarew / @jonahwilliams thoughts on using the temporary workaround for the time being?

Reply:
Can you refresh my memory on this @grouma? :)

Reply:
Right now we don't restore breakpoints if you refresh the application. `package:dwds` can handle refreshing the breakpoints through this deprecated [argument](https://github.com/dart-lang/webdev/blob/df632ec25bd68039a28b9624fa7a4dc4148a5649/dwds/lib/dwds.dart#L91). However, `webdev` nor the Flutter CLI make use of this functionality. It's only used by internal tools.

Reply:
Ah okay, gotcha. So, this is what the [requirePermissionToResume](https://pub.dev/documentation/vm_service/latest/vm_service/VmService/requirePermissionToResume.html) RPC is for. Clients are responsible for resetting their breakpoints after reloads, so the `requirePermissionToResume` ensures that an isolate stays paused until all clients from whom the service requires permission have indicated the isolate can be resumed.